panic on graceful shutdown when using webhook poller (close of closed channel)

hi,
while running a bot in webhook mode we consistently hit a panic during shutdown:
```panic: close of closed channel
goroutine ...:
gopkg.in/telebot.v3.(*Webhook).waitForStop(...)
    webhook.go:155
gopkg.in/telebot.v3.(*Webhook).Poll.func1(...)
    webhook.go:142 +0x2f
```

it looks like waitForStop closes the same stop channel that is already closed by the caller, leading to a double-close.
steps to reproduce
create a bot with a basic webhook setup:
```pref := tele.Settings{
    Token: "TOKEN",
    ParseMode: tele.ModeMarkdown,
}
b, _ := tele.NewBot(pref)
wh := &tele.Webhook{
    Listen:   ":8080",
    Endpoint: &tele.WebhookEndpoint{PublicURL: "https://example.com"},
}
b.Poller = wh
go b.Start()

// later…
b.Stop()
```


start the program and terminate it with ctrl-c (or call Stop() in code).
observe the panic.
expected
Bot.Stop() should shut the poller down cleanly without panicking.
actual
a runtime panic is raised because the channel is closed twice.
environment
telebot v3.1.0
go 1.22.0
os: linux (wsl2)
possible fix
ensure waitForStop returns without closing the channel, or gate the close with a select/sync.Once to avoid double closing.
thanks for looking into this – happy to test a patch or provide more details if needed.

got the same:

```
panic: close of closed channel

goroutine 6 [running]:
gopkg.in/telebot%2ev4.(*Webhook).waitForStop(0x140000e6370, 0x140003220e0)
	/Users/ksusonic/go/pkg/mod/gopkg.in/telebot.v4@v4.0.0-beta.5/webhook.go:161 +0x34
gopkg.in/telebot%2ev4.(*Webhook).Poll.func1(0x140003220e0)
	/Users/ksusonic/go/pkg/mod/gopkg.in/telebot.v4@v4.0.0-beta.5/webhook.go:148 +0x38
created by gopkg.in/telebot%2ev4.(*Webhook).Poll in goroutine 5
	/Users/ksusonic/go/pkg/mod/gopkg.in/telebot.v4@v4.0.0-beta.5/webhook.go:147 +0x340
```

--- a/webhook.go
+++ b/webhook.go
@@ -39,7 +39,6 @@ type WebhookEndpoint struct {
 // add the Webhook to a http-mux.
 //
 // If you want to ignore the automatic setWebhook call, you can set IgnoreSetWebhook to true.
-//
 type Webhook struct {
 	Listen           string   `json:"url"`
 	MaxConnections   int      `json:"max_connections"`
@@ -135,7 +134,7 @@ func (h *Webhook) Poll(b *Bot, dest chan Update, stop chan struct{}) {
 	h.bot = b
 
 	if h.Listen == "" {
-		h.waitForStop(stop)
+		<-stop
 		return
 	}
 
@@ -145,7 +144,7 @@ func (h *Webhook) Poll(b *Bot, dest chan Update, stop chan struct{}) {
 	}
 
 	go func(stop chan struct{}) {
-		h.waitForStop(stop)
+		<-stop
 		s.Shutdown(context.Background())
 	}(stop)
 
@@ -154,11 +153,6 @@ func (h *Webhook) Poll(b *Bot, dest chan Update, stop chan struct{}) {
 	} else {
 		s.ListenAndServe()
 	}
-}
-
-func (h *Webhook) waitForStop(stop chan struct{}) {
-	<-stop
-	close(stop)
 }
 
 // The handler simply reads the update from the body of the requests

--- a/webhook_test.go
+++ b/webhook_test.go
@@ -1,0 +1,160 @@
+package telebot
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http/httptest"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestWebhook_Poll(t *testing.T) {
+	t.Run("graceful shutdown", func(t *testing.T) {
+		webhook := &Webhook{
+			Listen:           "127.0.0.1:80",
+			IgnoreSetWebhook: true,
+		}
+
+		pref := defaultSettings()
+		pref.Offline = true
+		pref.Poller = webhook
+
+		b, err := NewBot(pref)
+		require.NoError(t, err)
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+			b.Start()
+		}()
+
+		// Give bot time to start
+		time.Sleep(100 * time.Millisecond)
+
+		b.Stop()
+		wg.Wait()
+	})
+
+	t.Run("no Listen", func(t *testing.T) {
+		webhook := &Webhook{
+			Listen:           "",
+			IgnoreSetWebhook: true,
+		}
+
+		pref := defaultSettings()
+		pref.Offline = true
+		pref.Poller = webhook
+
+		b, err := NewBot(pref)
+		require.NoError(t, err)
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+			b.Start()
+		}()
+
+		// Give bot time to start
+		time.Sleep(100 * time.Millisecond)
+
+		b.Stop()
+		wg.Wait()
+	})
+}
+
+func TestWebhook_ServeHTTP(t *testing.T) {
+	webhook := &Webhook{
+		IgnoreSetWebhook: true,
+	}
+
+	pref := defaultSettings()
+	pref.Offline = true
+	b, err := NewBot(pref)
+	require.NoError(t, err)
+
+	dest := make(chan Update, 1)
+	webhook.dest = dest
+	webhook.bot = b
+
+	t.Run("valid update", func(t *testing.T) {
+		update := Update{
+			ID: 123,
+			Message: &Message{
+				ID:   1,
+				Text: "test message",
+				Chat: &Chat{ID: 456},
+			},
+		}
+
+		body, _ := json.Marshal(update)
+		req := httptest.NewRequest("POST", "/", bytes.NewReader(body))
+		w := httptest.NewRecorder()
+
+		webhook.ServeHTTP(w, req)
+
+		select {
+		case received := <-dest:
+			assert.Equal(t, update.ID, received.ID)
+			assert.Equal(t, update.Message.Text, received.Message.Text)
+		case <-time.After(time.Second):
+			t.Fatal("timeout waiting for update")
+		}
+	})
+
+	t.Run("invalid JSON", func(t *testing.T) {
+		req := httptest.NewRequest("POST", "/", bytes.NewReader([]byte("invalid json")))
+		w := httptest.NewRecorder()
+
+		webhook.ServeHTTP(w, req)
+
+		select {
+		case <-dest:
+			t.Fatal("should not receive update for invalid JSON")
+		case <-time.After(100 * time.Millisecond):
+			// Expected - no update should be sent
+		}
+	})
+
+	t.Run("secret token validation", func(t *testing.T) {
+		webhook.SecretToken = "secret123"
+
+		update := Update{ID: 123}
+		body, _ := json.Marshal(update)
+
+		// Without secret token
+		req := httptest.NewRequest("POST", "/", bytes.NewReader(body))
+		w := httptest.NewRecorder()
+		webhook.ServeHTTP(w, req)
+
+		select {
+		case <-dest:
+			t.Fatal("should not receive update without secret token")
+		case <-time.After(100 * time.Millisecond):
+			// Expected
+		}
+
+		// With correct secret token
+		req = httptest.NewRequest("POST", "/", bytes.NewReader(body))
+		req.Header.Set("X-Telegram-Bot-Api-Secret-Token", "secret123")
+		w = httptest.NewRecorder()
+		webhook.ServeHTTP(w, req)
+
+		select {
+		case received := <-dest:
+			assert.Equal(t, update.ID, received.ID)
+		case <-time.After(time.Second):
+			t.Fatal("timeout waiting for update")
+		}
+
+		// Reset for other tests
+		webhook.SecretToken = ""
+	})
+}

